luarules/gadgets/unit_builder_priority.lua

```diff
@@ -82,207 +177,427 @@ local updateFrame = {}
+				-- We still have resources left over after completing the first pass, so do a second pass too 
+				if firstpass then 
+					roundRobinIndex = 1
 				end
 			end
 		end
+	else
+	-- Special case, we are completely and totally stalled, no resources left for passive builders. 
 
-		-- turn this passive builder on/off as appropriate
-		local wantedBuildSpeed = (wouldStall or not passiveConsExpense[builderID]) and 0 or realBuildSpeed[builderID]
-		if currentBuildSpeed[builderID] ~= wantedBuildSpeed then
-			spSetUnitBuildSpeed(builderID, wantedBuildSpeed)
-			currentBuildSpeed[builderID] = wantedBuildSpeed
-		end
-
-		-- override buildTargetOwners build speeds for a single frame; let them build at a tiny rate to prevent nanoframes from possibly decaying
-		if buildTargetOwners[builderID] and currentBuildSpeed[builderID] == 0 then
-			spSetUnitBuildSpeed(builderID, 0.001) --(*)
+		for builderID in pairs(passiveConsTeam) do 
+			if passiveExpense[builderID] and not midPrioSolarMaker[builderID] then  
+				MaybeSetWantedBuildSpeed(builderID, 0)
+			end
+		end	
+	end
+	
+	-- override buildTarget builders build speeds for a single frame; let them build at a tiny rate to prevent nanoframes from possibly decaying (yes this is confirmed to happen)
+	-- dont remove the resources given to a builder in a round-robin fashion just because they are build target owners
+	-- and take them off the buildTargets queue!
+	for buildTargetID, builderUnitID in pairs(buildTargets) do 
+		-- if owner is passive, then give it a bit of BP
+		-- This ensures that we at least build each unit a tiny bit.
+		if currentBuildSpeed[builderUnitID] < buildPowerMinimum then 
+			-- This builderUnitID has been assigned as passive with no resources, so give it a little bit
+			MaybeSetWantedBuildSpeed(builderUnitID, buildPowerMinimum)
+		else
+			-- this builderUnitID has been assigned greater than 0 resources in the round robin pass, so remove it from the next frames clear pass
+			buildTargets[buildTargetID] = nil
 		end
+	end 
+	
+	
+	Spring.SetTeamRulesParam(teamID, totalBuildPowerRule, totalBuildPower)
+	Spring.SetTeamRulesParam(teamID, highPrioBuildPowerRule, highPrioBuildPower)
+	Spring.SetTeamRulesParam(teamID, lowPrioBuildPowerRule, lowPrioBuildPower)
+	Spring.SetTeamRulesParam(teamID, highPrioBuildPowerWantedRule, highPrioBuildPowerWanted)
+	Spring.SetTeamRulesParam(teamID, lowPrioBuildPowerWantedRule, lowPrioBuildPowerWanted)
+	Spring.SetTeamRulesParam(teamID, highPrioBuildPowerUsedRule, highPrioBuildPowerUsed)
+	Spring.SetTeamRulesParam(teamID, lowPrioBuildPowerUsedRule, lowPrioBuildPowerUsed)
+	
+	if VERBOSE then 
+		Spring.Echo(string.format("%d Pstart = %.1f/%.0f Pleft = %.1f/%.0f RRI=%d #passive=%d Stalled=%d", 
+			Spring.GetGameFrame(),
+			passiveMetalStart, passiveEnergyStart,
+			passiveMetalLeft, passiveEnergyLeft, 
+			roundRobinIndex, 
+			numPassiveCons,
+			havePassiveResourcesLeft and 0 or 1
+			)
+		)
 	end
 end
 
-
 local function GetUpdateInterval(teamID)
```

Comment:
I know it's not changing in this PR, but I believe this logic is backwards: it starts maxInterval at 1 and increases it based on the resource least likely to overflow, when it should instead start at 6 and be reduced (for more frequent updates) if either resource requires it. I've fixed that in this commit: https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2600/commits/91bb4b285a7a3afd18c94cf422176a38ed533e6b

Reply:
Mmm, I should expose updaterate as a teamrulesparam and show that in debug view.

Reply:
Exposed it via teamrulesparam